### DISCUSSION TITLE
Compute shader: support counter resource?

### DISCUSSION THREAD

**Discussion by WangHoi** [DiligentGraphics/DiligentEngine#309] [Ideas]
Hello,  AppendStructuredBuffer/ConsumeStructuredBuffer seems not working now.

d3d12 device always create UAV with null counter resource.

**Comment by TheMostDiligent** [comment]
AppendStructuredBuffer/ConsumeStructuredBuffer is simply a buffer with a hidden counter in D3D. No other graphics APIs have the notion of counter buffers and emulating them in Diligent will only add complexity without any value as they can be straightforwardly implemented as structured buffer + counter buffer. Moreover, the explicit approach will be much more flexible as the app will have direct access to the counter.

**Comment by WangHoi** [comment]
Thanks, I'll port the code to use `RWByteAddressBuffer.InterlockedAdd`.
It should be slower than UAV counter.

**Reply by TheMostDiligent** [reply]
Why would it be slower?

**Reply by WangHoi** [reply]
AFAIK, Some gpu provides a very fast on-chip globally coherent cache called `GDS`.
[AMD’s Cayman GPU Architecture](https://www.realworldtech.com/cayman/7/)

So beyond `RWByteAddressBuffer.InterlockedAdd`, D3D11 and D3D12 provides the UAV counter api, to expose this hardware ability.

**Reply by TheMostDiligent** [reply]
Accessing a single counter from the entire GPU should be done carefully no matter how it is implemented internally.
If your compute shader needs to write 10,000 elements, you'd better off incrementing the global counter once by 10,000 to reserve space for all elements, and then writing the elements within the compute group.
Adding 10,000 elements one-by-one is extremely unlikely to work better even if the counter resides in GDS.

**Reply by MikhailGorobets** [reply]
Tera Scale is a very old architecture, so this article cannot be taken seriously at the moment. Try writing a benchmark that uses AppendBuffer, and another one that uses atomic operations with InterlockedAdd, and we’ll see if this can actually affect performance.
